### PR TITLE
fuzz: Faster utxo_snapshot fuzz target

### PR DESCRIPTION
**Pull request by maflcko**
Two commits to speed up unit and fuzz tests.

Can be tested by running the fuzz target and looking at the time it took, or by looking at the flamegraph. For example:

```
FUZZ=utxo_snapshot perf record -g --call-graph dwarf ./src/test/fuzz/fuzz  -runs=100
hotspot ./perf.data

### CODE REVIEW

**Comment by dergoegge** [comment]
What is the expected speed up?

**Review by sedited — APPROVED** [review]
Nice, ACK fa48ded21168740640ff9d6e5b413a322d58b114

I'd re-ack quickly if you want to fix the other includes as well.

**Review by marcofleon — COMMENTED** [review]
Tested ACK fa198d1a196482bf363ba57a3f4d6d2d860b8ce5

**Comment by marcofleon** [comment]
Sounds good, I'll review again when it's improved then. Looked to be about 2x when I just tested it.

**Review by sedited — COMMENTED** [review]
I still see a doubling in performance of the valid test if the validation cache is initialized with minimal values. Do you want to add it back?

current HEAD:
```
#10000	DONE   cov: 2794 ft: 3029 corp: 66/642b lim: 33 exec/s: 161 rss: 406Mb
Done 10000 runs in 62 second(s)
```
`+` min validation cache:
```
#10000	DONE   cov: 2788 ft: 3011 corp: 54/431b lim: 68 exec/s: 333 rss: 350Mb
Done 10000 runs in 30 second(s)
```

File: src/test/fuzz/utxo_snapshot.cpp [hunk 1]
```diff
@@ -1,45 +1,79 @@
...
 #include <validation.h>
-#include <validationinterface.h>
+
+#include <cstdint>
+#include <functional>
+#include <ios>
+#include <memory>
+#include <optional>
+#include <ranges>
+#include <vector>
```
Comment (TheCharlatan) [inline src/test/fuzz/utxo_snapshot.cpp:35]: Nit: The includes need to be corrected again.

**Review by sedited — APPROVED** [review]
ACK fa69afd51879f8441df7a51144318a49e7dd27b9

The logic is a bit convoluted, but I don't have many suggestions for improving it and I think it is important to get this performance improvement deployed quickly.

**Comment by marcofleon** [comment]
exec:s
exec:s_dos

The second one just shows the same plot zoomed in on the last 600k iterations or so. I'm going to follow up with a plot of executions vs time as well. And I'll do more than a million iterations, as it seems like `utxo_snapshot_invalid` might end up being faster if fuzzed for hours.

I compiled using  only `--with-sanitizers=fuzzer`. Should I test for speed using other sanitizers or will the performance difference be similar regardless?

**Comment by marcofleon** [comment]
Thank you for the suggestions, I'll do more trials for sure. That actually is the `utxo_snapshot_invalid` target compared to the old `utxo_snapshot` (sorry, file names should've been clearer). It's much faster at the beginning. I ran it from scratch, not on any corpus. I'm generating a corpus now for the invalid target and then I'll do a comparison for both when fuzzing from a corpus. 

Also, for the bug from https://github.com/bitcoin/bitcoin/issues/30514#issuecomment-2285605841. What's the reason for `max_len=128`? Is that necessary to find the bug in a reasonable amount of time?

**Comment by marcofleon** [comment]
To clarify, I reproduced the bug fairly quickly with the line you provided in https://github.com/bitcoin/bitcoin/pull/30644#issuecomment-2290858777. I would like to see what happens with `max_len` not specified.

**Comment by marcofleon** [comment]
I'll recheck my setup and get back to you.

**Review by sedited — APPROVED** [review]
Re-ACK fa0b66400c12d11579b38b1ac76f677b23e8c82a

I saw the old "libFuzzer disabled leak detection" info message in the invalid test, but not sure what might be causing it.

**Comment by marcofleon** [comment]
I'm using Debian. Here's my recent run (2 million iterations) using an initial corpus for both targets:

Screenshot 2024-08-19 at 10 54 46 AM

The new target is clearly faster. I think this is a more useful run because we generally fuzz from a corpus of inputs. I also reproduced the bug using only `-use_value_profile=1`. So overall, I would say this PR is good to go. I agree that it's probably not really worth spending time to improve it for non-ramdisk fuzzing, but up to you.

It is interesting that the new target eventually slows down significantly when fuzzing from no inputs. I ran this:
`FUZZ=utxo_snapshot_invalid perf record -g --call-graph dwarf ./src/test/fuzz/fuzz  -runs=200000`
and here is flamegraph: https://marcofleon.github.io/flamegraph/19-8-utxo_snap_invalid.svg

**Comment by marcofleon** [comment]
Let me know if you think there is anything more I can do to test. Looking at that flame graph is fun so I'm happy to keep fuzzing.

**Comment by marcofleon** [comment]
Makes sense, was just working with Niklas to get this done. The flame graphs are a useful new tool for me too, I'll be utilizing those more. Thanks for the help @maflcko.

**Comment by marcofleon** [comment]
ACK fa0b66400c12d11579b38b1ac76f677b23e8c82a

File: src/test/util/setup_common.h [hunk 2]
```diff
@@ -28,7 +28,6 @@ class arith_uint256;
 class CFeeRate;
 class Chainstate;
 class FastRandomContext;
-
```
Comment (TheCharlatan) [inline src/test/util/setup_common.h:31]: I don't think this empty line should be removed.

**Review by sedited — APPROVED** [review]
Re-ACK fa899fb7aa8a14acecadd8936ad5824fa0f697ff

**Review by marcofleon — APPROVED** [review]
Re ACK fa899fb7aa8a14acecadd8936ad5824fa0f697ff

**Review by brunoerg — APPROVED** [review]
ACK fa899fb7aa8a14acecadd8936ad5824fa0f697ff

Left it running for a while, got a clear improvement compared to master.

**Pull request merged**